### PR TITLE
Blog Post: Setting Django model fields automatically using Oracle Database sequences

### PR DESCRIPTION
**Pull request by merit-mthompson**
Details how to use `set_sql_for_field` for grabbing sequences or other arbitrary bits from the database for a model save. Also updating some styles and docs along with upgrading core stuff for security.

Full post in context:

Screenshot_2020-11-11 Setting Django model fields automatically using Oracle Database sequences

### CODE REVIEW

**Comment by merit-ddennis** [comment]
This post looks well-constructed. How do I review/approve?

File: content/blog/django-model-fields-oracle-database-sequences/index.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+Use the `set_sql_for_field` decorator below on your model's `save()` method to get/set a sequence or any other arbitrary SQL.
+
+# An imperfect database for an imperfect world
+
+Sometimes when working with existing or legacy databases you might run into less-than-perfect implementations of things like primary keys.
+
+Could be that when the database was created the database management system (DBMS) didn't support things like `AUTO_INCREMENT` (MySQL) or `IDENTITY` (Oracle Database 12+). It could also be that those responsible had good reason to manually manage things or maybe didn't know of a different way of doing things.
+
+**Let's say for reasons beyond this blog post that you cannot alter the existing datbases/tables.**
```
Comment (merit-rluk) [inline content/blog/django-model-fields-oracle-database-sequences/index.md:23]: s/datbases/databases/

**Review by merit-rluk — APPROVED** [review]
Looks great!

**Review by merit-dsclair — APPROVED** [review]
Love this! Looks great!

File: content/blog/django-model-fields-oracle-database-sequences/index.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+class DepartmentTypeSerializer(ModelSerializer):
+    class Meta:
+        fields = '__all__'
+        model = DepartmentType
+        read_only_fields = ['type_key']
+```
+
+# Closing thoughts
+
+**Whenever possible use primary keys and `AUTO_INCREMENT` or `IDENTITY`**. For the times you can't, hopefully something like `set_sql_for_field` can help!
```
Comment (schelcj) [inline content/blog/django-model-fields-oracle-database-sequences/index.md:221]: Why? If anything you're making the case that those aren't needed. There are valid reasons for using sequences, just googling sequences vs auto_increment will show lots of debate, and there is even the case for using uuid generation in postgres (which btw, your decorator will handle wonderfully.) We should probably backup a recommendation like this somehow.

**Review by merit-jrussell — APPROVED** [review]
I can't comment on the correctness of the code, but everything else looks good as-is.

File: content/blog/django-model-fields-oracle-database-sequences/index.md (hunk 1 above)
Comment (merit-mthompson) [reply content/blog/django-model-fields-oracle-database-sequences/index.md]: Fixed, thanks for catching! Ran a full spell check in NeoVim, just to be sure I didn't miss anything else.

File: content/blog/django-model-fields-oracle-database-sequences/index.md (hunk 2 above)
Comment (merit-mthompson) [reply content/blog/django-model-fields-oracle-database-sequences/index.md:221]: In Oracle `IDENTITY` is using sequences under the hood and we're not saying not to use sequences. I'm not sure I follow how we're making the case that they're not needed. The article is about something you can't change that maybe wasn't built in the best way and how to work around that.

The recommendation is to use `IDENTITY` as a convenience and to get some of the extra features/requirements for identity like only being able to use one of them on a table (more at https://docs.oracle.com/database/121/SQLRF/statements_7002.htm#SQLRF55661).

The intent of the wrap-up isn't to consider every possible way to assign an ID in every DBMS but to give a high-level "use the magic" suggestion.

Comment (schelcj) [reply content/blog/django-model-fields-oracle-database-sequences/index.md:221]: > The article is about something you can't change that maybe wasn't built in the best way and how to work around that.

Then maybe leave the recommendation off entirely?

Comment (merit-mthompson) [reply content/blog/django-model-fields-oracle-database-sequences/index.md]: We're trying to educate and I don't agree that "use primary keys and IDs" is in any way controversial. I also think saying nothing isn't helpful at all.

We've spent a good bit of the article effectively saying, "the way it was built wasn't ideal" and it makes sense to close with a "here's how to do it right".

I've updated the PR, let me know if this feels a little better.

Comment (schelcj) [reply content/blog/django-model-fields-oracle-database-sequences/index.md]: Nothing here is controversial. The article is about dealing with legacy schema and that's great but that message is mixed with here's the way to do it right without anything backing up why we feel that's the right approach.

Comment (merit-mthompson) [reply content/blog/django-model-fields-oracle-database-sequences/index.md]: If it's not controversial to suggest auto-incrementing primary keys as a sane default I'm not sure why we need to back it up. To me, that's been a longstanding best-practice.

Are you hoping to see links or something to why you should (generally) use them?

**Review by schelcj — APPROVED** [review]

File: content/blog/django-model-fields-oracle-database-sequences/index.md (hunk 2 above)
Comment (merit-mthompson) [reply content/blog/django-model-fields-oracle-database-sequences/index.md]: We've talked a bit about this via Slack and we're cool with how it is. We might write another article with more detail about how to do things the right way.

**Review by merit-ddennis — APPROVED** [review]
Great to see this ready to go. Appreciate the 'editorial debate' - I think we should remember that it's okay for any of these postings to be thought-provoking or challenging... if a reader wants to question what we're saying, then they should reach out to us. Let's keep it easy and fun, and our style can evolve over time. Feel free to publish!

**Pull request merged**